sled-agent/src/instance.rs

```diff
@@ -748,11 +749,22 @@ impl InstanceRunner {
             info!(self.log, "instance runner exited main loop");
         }
 
-        info!(self.log, "before publish_state_to_nexus");
         if state_owner == VmmStateOwner::Runner {
             self.publish_state_to_nexus().await;
         }
-        info!(self.log, "after publish_state_to_nexus");
+
+        // Drop the instance ticket so that this instance will be removed from
+        // the instance manager. This ensures that no more requests will be
+        // queued to the runner, which allows this routine to send replies to
+        // those requests in an orderly fashion.
+        //
+        // This needs to be done only after publishing the final VMM state to
+        // Nexus so that Nexus is guaranteed to have received notice of the
+        // VMM's terminal state before it begins to see "no such VMM" errors
+        // from the instance manager. Otherwise it is possible for Nexus to
+        // conclude that a cleanly-stopped instance has gone missing and should
+        // be restarted.
+        drop(ticket);
```

Comment:
This somewhat makes me wonder if tying this to dropping the ticket was a mistake, but ... yeah, seems reasonable

Reply:
Yeah... I think the important property here is that the ticket is no longer a member of `InstanceRunner`, which makes it much harder for `run`'s callees to call `deregister` when they shouldn't. I also sort of prefer the aesthetics of an explicit `deregister` routine if we're not going to rely on auto-drop, so I'll see about switching this back.